### DISCUSSION TITLE
do_string/file in GDScript

### DISCUSSION THREAD

**Discussion by JekSun97** [gilzoide/godot-lua-pluginscript#51] [Q&A]
is it possible to read lua code from gdscript? I'm making a programmer's simulator where you need to write code in a game in lua

**Comment by gilzoide** [comment]
Hi @JekSun97 !

> is it possible to read lua code from gdscript?

Well, not out of the box, Lua functions are only directly callable from Lua scripts.

But it's easy enough to make a `do_string` and `do_file` method that is callable from GDScript: just create a Lua script that has that functionality and register it as a [Singleton (Autoload)](https://docs.godotengine.org/en/3.0/getting_started/step_by_step/singletons_autoload.html) in Godot. A script such as the following should work:

```lua
local LuaRunner = {
  extends = 'Node'
}

function LuaRunner:do_string(code)
  -- using `tostring` here is necessary to convert Godot String -> Lua string
  local func = assert(load(tostring(code)))
  return func()
end

function LuaRunner:do_file(code)
  -- using `tostring` here is necessary to convert Godot String -> Lua string
  local func = assert(loadfile(tostring(code)))
  return func()
end

return LuaRunner
```

If you register this `LuaRunner` node as an Autoload, for example with the name `LuaRunner`, you can then access it from GDScript and call its methods:
```gdscript
extends Node

func _ready():
	LuaRunner.do_string("print 'Hello from Lua!'")
```

I haven't tested this code, but I think it will work fine =]

If can also sandbox the environment of the loaded code, to avoid user code crashing your game, accessing the file system, this stuff that malicious code might attempt to do. If it is just for an experiment/for fun, it's ok to overlook this, but you should most definitely worry about sandboxing if shipping this product to end users.
